### DISCUSSION TITLE
[Education] Alternative programming interfaces for py5 - nodes/wires or blocks

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#478] [Ideas]
@hx2A has shown me some very cool  ["nodes and wires interfaces"](https://github.com/xyflow/awesome-node-based-uis) that could be interesting to explore, like https://ryven.org/ and https://wonderworks-software.github.io/PyFlow/, but nodes are kind of not my cup of tea.

I'm trying to convince some friends that like JavaScript to see if they want to make an offline (local, not online) "Python with blocks" tool based on https://developers.google.com/blockly maybe. Inspired by (onlline) https://trinket.io/blocks or https://oldapp.edublocks.org/july23/project/new?mode=python&type=blocks, that could integrate with a full CPython interpreter (and py5).

**Comment by hx2A** [comment]
I'm curious about these possibilities, but don't see myself actually using it much, which makes it harder to imagine what this could or would look like. I don't know much about visual programming. If py5 had a way for doing visual programming, what exactly would it look like? What kind of nodes would there be? Maybe we should start by drawing a node diagram to help flush out ideas here.

**Reply by villares** [reply]
For the nodes bit, I think we should convince people who use and like nodes, like @guilhermesv, to give us some ideas.

**Reply by hx2A** [reply]
What would make something like this really useful if there was a "toolbox" of nodes one could use to quickly do more sophisticated things. For example, right now I am doing something that would benefit from ability to pass colors through configurable color channel curves, much like the color channel curves available for editing photos in tools like Gimp or Krita. If there was a color channel node, it would be easy to link it into a program.

**Reply by guilhermesv** [reply]
Hello, I really like the operating logic and functions of NodeBox 3, I think it's a good initial reference.

It follows this idea of ​​having nodes to do more sophisticated things like grid, disperse points randomly in a shape, find a point on a line, generate a wave. But it also has basic things like simple mathematical operations.

Here is a list of available nodes:
https://www.nodebox.net/node/reference/

It's not very big, but it allows a lot of things.

For the interface, I think that in this type it is cool to have real-time feedback and also to have an easy way to export the result as an image or video.

**Reply by hx2A** [reply]
Thank you, @guilhermesv , nodebox 3 looks very interesting. I need to experiment with this to generate ideas for how this can be used with or for py5.

**Reply by Maltiec** [reply]
also Nodebox 3 allow to write user defined nodes in python (is this option is still exists)

**Comment by villares** [comment]
While I don't have at the moment the time, skill or energy to pursue an offline "Python with blocks" interface, I have been slowly working on my kind-of-block based Parsons puzzels...  https://abav.lugaralgum.com/faded-parsons-visual/ (more info at: https://github.com/villares/faded-parsons-visual)
![image](https://github.com/user-attachments/assets/266ac9da-9a54-4760-86d6-7e929a8fa923)
![image](https://github.com/user-attachments/assets/967ab171-04bd-4a03-a649-e5f264ec9aaf)
